daemon/logger/ring.go

```diff
@@ -0,0 +1,213 @@
+package logger
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+)
+
+const (
+	defaultRingMaxSize = 1000000 // 1MB
+)
+
+// RingLogger is a ring buffer that implements the Logger interface.
+// This is used when lossy logging is OK.
+type RingLogger struct {
+	buffer *messageRing
+	ctx    context.Context
+	cancel func()
+	l      Logger
+	logCtx Context
+}
+
+type ringWithReader struct {
+	*RingLogger
+}
+
+func (r *ringWithReader) ReadLogs(cfg ReadConfig) *LogWatcher {
+	reader, ok := r.l.(LogReader)
+	if !ok {
+		// something is wrong if we get here
+		panic("expected log reader")
+	}
+	return reader.ReadLogs(cfg)
+}
+
+func newRingLogger(driver Logger, logCtx Context, maxSize int64) *RingLogger {
+	ctx, cancel := context.WithCancel(context.Background())
+	l := &RingLogger{
+		buffer: newRing(maxSize),
+		ctx:    ctx,
+		cancel: cancel,
+		l:      driver,
+		logCtx: logCtx,
+	}
+	l.run()
+	return l
+}
+
+// NewRingLogger creates a new Logger that is implemented as a RingBuffer wrapping
+// the passed in logger.
+func NewRingLogger(driver Logger, logCtx Context, maxSize int64) Logger {
+	if maxSize < 0 {
+		maxSize = defaultRingMaxSize
```

Comment:
Yes, but I wanted to support a 0 size, which would essentially mean store no more than 1 message.